load.c

```diff
@@ -689,6 +689,19 @@ rb_provide(const char *feature)
 
 NORETURN(static void load_failed(VALUE));
 
+static inline VALUE
+realpath_internal_cached(VALUE hash, VALUE path)
+{
+    VALUE ret = rb_hash_aref(hash, path);
+    if(RTEST(ret)) {
+        return ret;
+    }
+
+    VALUE realpath = rb_realpath_internal(Qnil, path, 1);
+    rb_hash_aset(hash, rb_fstring(path), rb_fstring(realpath));
+    return realpath;
+}
```

Comment:
Just an idea: this function accesses the hash for the same element twice.

```suggestion
static int
realpath_cache_update(st_data_t *path, st_data_t *real, st_data_t arg, int existing)
{
    if (!existing) {
        VALUE realpath = rb_realpath_internal(Qnil, *path, 1);
        *real = rb_fstring(realpath);
    }
    *(VALUE *)arg = *real;
    return ST_STOP;
}

static inline VALUE
realpath_internal_cached(VALUE hash, VALUE path)
{
    VALUE ret;
    rb_hash_stlike_update(hash, (st_data_t)path, realpath_cache_update, (st_data_t)&ret);
    return ret;
}
```

You'll need to include `"internal/hash.h"` and update dependencies in common.mk too.

Reply:
Hi @nobu, I have tested your suggestion and it seems to add  `1.3 ms` to the execution time on my environment. Is it possible to merge without this change? Thanks.